lib/plugins/filter/meta_generator.js

```diff
@@ -1,21 +1,14 @@
 'use strict';
 
-let cheerio;
-const hexoGeneratorTag = '<meta name="generator" content="Hexo %s" />';
-
 function hexoMetaGeneratorInject(data) {
   const { config } = this;
-  if (!config.meta_generator) return;
-
-  if (!cheerio) cheerio = require('cheerio');
-  const $ = cheerio.load(data, {decodeEntities: false});
+  if (!config.meta_generator ||
+  data.includes('<meta name="generator" content="Hexo')) return;
 
-  if (!($('meta[name="generator"]').length > 0) &&
-  $('head').contents().length > 0) {
-    $('head').prepend(hexoGeneratorTag.replace('%s', this.version));
+  const hexoGeneratorTag = '<meta name="generator" content="Hexo %s" />'
+    .replace('%s', this.version);
 
-    return $.html();
-  }
+  return data.replace(/<\/head>/, hexoGeneratorTag.concat('</head>'));
```

Comment:
Without `/g` global modifier, only the first occurrence is replaced. I tested with cheerio@1.0.0-rc.3, `<head></head>` is only found in body, so this approach is compatible.

A safer way to do this is to avoid `<head></head>` using regex. This is so that it can work in an edge case like,
``` html
<head>
<link>
<head></head>
</head>
```

The regex I came up so far is `/(?!<head>)<\/head>/g`, but it doesn't work.

``` js
const headTag = '<head><span></span></head><head></head>'

console.log(headTag.replace(/(?!<head>)<\/head>/g, 'ccc</head>'))
// <head><span></span>ccc</head><head>ccc</head>
// expected <head><span></span>ccc</head><head></head>
```

Reply:
If without `g`, this regex will only match the first `</head>` after `<head>`.

```javascript
const headTag = '<head><span></span></head><head></head>'
console.log(headTag.replace(/(?!<head>)<\/head>/, 'ccc</head>'))
//  <head><span></span>ccc</head><head></head>
```

However, there might be some other problems, in case:

```ejs
// layout.ejs
<!DOCTYPE html>
<!--<html><head></head></html>-->
<html <% if(config.language) { %>lang="<%- config.language %>"<% } %>>
```

Reply:
Somehow `/(?!<head>)<\/head>/g` is equivalent to `/<\/head>/g`.
`/(?!<head>)<\/head>/g` matches every `</head>` regardless whether it's preceded by `<head>`;

tested on [regex101.com](https://regex101.com/r/QEfJTB/1) and [regexr.com](https://regexr.com/), same thing.

Reply:
`<head>(?!<\/head>)`

Check this out? This one can match every `<head>` which doesn't follow `</head>`

Reply:
Thanks for the suggestion, but I prefer not to mess with `<meta charset="utf-8">` which is usually after `<head>`. W3C [recommends](https://www.w3.org/International/questions/qa-html-encoding-declarations) that way.

> The declaration should fit completely within the first 1024 bytes at the start of the file, so it's best to put it immediately after the opening head tag.

Reply:
Then situation becomes very complicated, because theme might not include both of `meta[generator]` and `meta[charset]`.

Reply:
What about before/after `<title>? ? I assume theme is more likely to have <title> than <meta charset>.

Reply:
@curbengh I think it is a good idea!

We can first detect `meta[generator]`, if none of them detected, we can then replace `</title>` with `</title><meta generator="">`

Reply:
I'll add it after `</title>`. hexo-theme-landscape [inserts meta tags](https://github.com/hexojs/hexo-theme-landscape/blob/master/layout/_partial/head.ejs) after title. I assume most theme follow that convention.

Reply:
I have done some small research on `<head>` contents before. `<title>` should come as early as possible, right after `charset` `preload` `prefetch`, so that browser can fill the title in the tab immediately.

Reply:
We can use `/<meta\s+name=['|"]?generator['|"]?/i` to detect if current HTML already includes `meta[generator]`.
Although it will match something like `<meta name="generator'>` (https://github.com/hexojs/hexo/pull/3129#discussion_r183027756), but since the match means `meta[generator]` has been added, I assumed the author of the theme won't make such a stupid error.

Reply:
Is regex necessary when `data.includes('<meta name="generator")` is suffice? I use `data.includes('<meta name="generator" content="Hexo')` simply to prevent it being added more than once.

If user prefer custom value, they can always disable meta_generator and add their version in the theme.

Reply:
I think detect `meta generator` is a better option. Maybe some theme will use `meta generator="xxx Hexo"`, thus we should detect `generator` attribute only.

Reply:
Added `/<meta\s+name=['|"]?generator['|"]?/i`.

Reply:
LGTM then. I will bring up a performance test to see if that regex have some impact on generating speed.

Reply:
https://travis-ci.com/theme-suka/performance-test/builds/123489483

**Test A**

latest hexo master branch with `meta_generator` set to false

![image](https://user-images.githubusercontent.com/40715044/63161120-e5b15980-c051-11e9-9d5f-29a2a9980579.png)

**Test B**

latest hexo curbengh:metagen-cheerio branch with `meta_generator` set to true

![image](https://user-images.githubusercontent.com/40715044/63161405-9f102f00-c052-11e9-81a6-cd293730fef6.png)

I think it is good to go.